### ISSUE TITLE
Serverless V4 esbuild not including files that are not mentioned in the serverless config like utils and helper files

### ISSUE THREAD

**Issue by visrut-at-handldigital** [serverless/serverless#12744]
### Issue description

I have started using V4, and it's going all well, my codebase is in javascript but since serverless v4 native javascript/typescript support with a little bit of configuration, I have started to convert a few files to typescript.

### **Issue:**

When I run the `serverless package` command, it doesn't include helper files like `util.ts` in my package or zip, and when I deploy it, I get no module found error.

### **Steps to reproduce:**

Created a repository with commands I have run in the README file: [https://github.com/visrut-at-handldigital/aws-serverless](<https://github.com/visrut-at-handldigital/aws-serverless>)

I read the documentation here: [https://www.serverless.com/framework/docs/providers/aws/guide/building](<https://www.serverless.com/framework/docs/providers/aws/guide/building>)<br>It's good but I didn't seem to find any option for `esbuild` that allows me to include other helper files.

**Note:**<br>The only reason I have kept `bundle: false` is that don't want to switch to bundling right now, I am thinking of it after I would convert all the files in TypeScript.

**Comment by visrut-at-handldigital** [comment]
Follow up:

Ok for the javascript part, I am good since I was able to include it inside the package zip by only mentioning in `packages.patterns` inside serverless.yml, but for typescript util and helper files, I would require a compilation step.

I was looking into `lib/plugins/esbuild/index.js` thinking if there is any option to configure it, but seems like not, I am happy to contribute as well but don't know much about esbuild but as far as I know maybe I have to add `include` option in **_defineSchema** method and then append new functions in **_build** method after getting functions from `await this.functions(handlerPropertyName)`. correct me if I am wrong or there is already way exist and don't need to do this.

**Comment by Mmarzex** [comment]
@visrut-at-handldigital thanks for reporting and we will look into it. If you set `bundle: true` does it work correctly?

**Comment by visrut-at-handldigital** [comment]
Yes `bundle: true` works fine, the only reason I am not using it now is that other code is in javascript and uses require.js and import files simple, so want to keep in that way until I migrate to fully typescript.

I can help to contribute as well, please let me know.

**Comment by casewalker** [comment]
Hello!

I have also come across this issue and my project decided to continue to use `serverless-plugin-typescript` as a result. With this usage, we get an NPM error on install:
```
Could not resolve dependency:
peer serverless@"2 || 3" from serverless-plugin-typescript@2.1.5
```

We are temporarily using `--force` to install the dependency anyway, but that isn't a great solution. We are working on migrating JS to TS throughout the project, but that may not be complete for some time.

Can you share if there has been any update on this issue?

And if it seems like this will not be resolved soon, would Serverless consider editing the `serverless-typescript-plugin` to allow <=4.4.X to be a peer dependency so that I do not have to use `--force` on my NPM installs?

**Comment by visrut-at-handldigital** [comment]
@casewalker I waited for updates for so long but then decided to completely migrate the stack to typescript first, even if I had to use the `any` type at first. I converted all the **require** statements to **import**, and all `.js` to `.ts` as it is without changing any logic other than giving types and import statements.

After I switched to mode `bundle=true` and now gradually removing all `any` types as I come across the code so far working well, I was also able to remove the `lambda-layers` we had which can increase the boot time.

I would suggest doing the same since earlier I thought about contributing as well but there is a blocker for that as well see this issue: https://github.com/serverless/serverless/issues/12761

**Comment by czubocha** [comment]
Released in [Serverless Framework 4.42.0](https://github.com/serverless/serverless/releases/tag/sf-core%404.42.0) via #13850. With `build.esbuild.bundle: false`, the service is packaged the way classic packaging works, with TypeScript compiled in place: every `.js`/`.ts` file in the service (respecting `package.patterns`) is included, so helpers such as `util.ts` ship with the handler. No configuration change is needed beyond upgrading. Details in [Building Without Bundling](https://www.serverless.com/framework/docs/providers/aws/guide/building#building-without-bundling).
